lib/semantics/check-io.cc

```diff
@@ -0,0 +1,608 @@
+  SpecifierKind specKind{};
+  using ParseKind = parser::InquireSpec::LogVar::Kind;
+  switch (std::get<parser::InquireSpec::LogVar::Kind>(spec.t)) {
+  case ParseKind::Exist: specKind = SpecifierKind::Exist; break;
+  case ParseKind::Named: specKind = SpecifierKind::Named; break;
+  case ParseKind::Opened: specKind = SpecifierKind::Opened; break;
+  case ParseKind::Pending: specKind = SpecifierKind::Pending; break;
+  }
+  SetSpecifier(specKind);
+}
+
+void IoChecker::Enter(const parser::IoControlSpec &spec) {
+  // IoControlSpec context Name
+  flags_.set(Flag::IoControlList);
+  if (std::holds_alternative<parser::Name>(spec.u)) {
+    SetSpecifier(SpecifierKind::Nml);
+    flags_.set(Flag::FmtOrNml);
+  }
+}
+
+void IoChecker::Enter(const parser::IoControlSpec::Asynchronous &spec) {
+  SetSpecifier(SpecifierKind::Asynchronous);
+  if (const std::optional<std::string> charConst{
+          GetConstExpr<std::string>(spec)}) {
+    flags_.set(
+        Flag::AsynchronousYes, parser::ToUpperCaseLetters(*charConst) == "YES");
+    CheckStringValue(SpecifierKind::Asynchronous, *charConst,
+        parser::FindSourceLocation(spec));  // C1223
+  }
+}
+
+void IoChecker::Enter(const parser::IoControlSpec::CharExpr &spec) {
+  SpecifierKind specKind{};
+  using ParseKind = parser::IoControlSpec::CharExpr::Kind;
+  switch (std::get<ParseKind>(spec.t)) {
+  case ParseKind::Advance: specKind = SpecifierKind::Advance; break;
+  case ParseKind::Blank: specKind = SpecifierKind::Blank; break;
+  case ParseKind::Decimal: specKind = SpecifierKind::Decimal; break;
+  case ParseKind::Delim: specKind = SpecifierKind::Delim; break;
+  case ParseKind::Pad: specKind = SpecifierKind::Pad; break;
+  case ParseKind::Round: specKind = SpecifierKind::Round; break;
+  case ParseKind::Sign: specKind = SpecifierKind::Sign; break;
+  }
+  SetSpecifier(specKind);
+  if (const std::optional<std::string> charConst{GetConstExpr<std::string>(
+          std::get<parser::ScalarDefaultCharExpr>(spec.t))}) {
+    if (specKind == SpecifierKind::Advance) {
+      flags_.set(
+          Flag::AdvanceYes, parser::ToUpperCaseLetters(*charConst) == "YES");
+    }
+    CheckStringValue(specKind, *charConst, parser::FindSourceLocation(spec));
+  }
+}
+
+void IoChecker::Enter(const parser::IoControlSpec::Pos &spec) {
+  SetSpecifier(SpecifierKind::Pos);
+}
+
+void IoChecker::Enter(const parser::IoControlSpec::Rec &spec) {
+  SetSpecifier(SpecifierKind::Rec);
+}
+
+void IoChecker::Enter(const parser::IoControlSpec::Size &spec) {
+  SetSpecifier(SpecifierKind::Size);
+}
+
+void IoChecker::Enter(const parser::IoUnit &spec) {
+  if (const parser::Variable * var{std::get_if<parser::Variable>(&spec.u)}) {
+    // TODO: C1201 - internal file variable must not be an array section ...
+    auto expr{GetExpr(*var)};
+    int defaultKind{
+        context_.defaultKinds().GetDefaultKind(TypeCategory::Character)};
+    if (expr && expr->GetType() && expr->GetType()->kind != defaultKind) {
```

Comment:
Using `ExprTypeKindIsDefault()` from `tools.h` might make this check simpler.

Reply:
Done